docs/rfcs/optional-config-type.md

```diff
@@ -0,0 +1,184 @@
+# Optional[T] type for use in configuration structs
+
+## Overview
+
+In Go, types are by default set to a "zero-value", a supported value for the
+respective type that is semantically equivalent or similar to "empty", but which
+is also a valid value for the type. For many config fields, the zero value is
+not a valid configuration value and can be taken to mean that the option is
+disabled, but in certain cases it can indicate a default value, necessitating a
+way to represent the presence of a value without using a valid value for the
+type.
+
+Using standard Go without inventing any new types, the two most straightforward
+ways to accomplish this are:
+
+1. Using a separate boolean field to indicate whether the field is enabled or
+   disabled.
+2. Making the type a pointer, which makes a `nil` pointer represent that a
+   value is not present, and a valid pointer represent the desired value.
+
+Each of these approaches has deficiencies: Using a separate boolean field
+requires the user to set the boolean field to `true` in addition to setting the
+actual config option, leading to suboptimal UX. Using a pointer value has a few
+drawbacks:
+
+1. It may not be immediately obvious to a new user that a pointer type indicates
+   a field is optional.
+2. The distinction between values that are conventionally pointers (e.g. gRPC
+   configs) and optional values is lost.
+3. Setting a default value for a pointer field when decoding will set the field
+   on the resulting config struct, and additional logic must be done to unset
+   the default if the user has not specified a value.
+4. The potential for null pointer exceptions is created.
+
+## Optional types
+
+Go does not include any form of Optional type in the standard library, but other
+popular languages like Rust and Java do. We could implement something similar in
+our config packages that allows us to address the downsides of using pointers
+for optional config fields.
+
+## Basic definition
+
+A production-grade implementation will not be discussed or shown here, but the
+basic implementation for an Optional type in Go could look something like:
+
+```golang
+type Optional[T any] struct {
+	hasValue bool
+	value    T
+
+	defaultVal T
+}
+
+func Some[T any](value T) Optional[T] {
+	return Optional[T]{value: value, hasValue: true}
+}
+
+func None[T any](value T) Optional[T] {
+	return Optional[T]{}
+}
+
+func WithDefault[T any](defaultVal T) Optional[T] {
+	return Optional[T]{defaultVal: defaultVal}
+}
+
+func (o Optional[T]) HasValue() bool {
+	return o.hasValue
+}
+
+func (o Optional[T]) Value() T {
+    return o.value
+}
+
+func (o Optional[T]) Default() T {
+    return o.defaultVal
+}
+```
+
+## Use cases
+
+Optional types can fulfill the following use cases we have when decoding config.
+
+### Representing optional config fields
```

Comment:
Could we list other use cases? e.g. TLSConfig

Reply:
Sure thing, I added another example for how this would impact `confighttp.ServerConfig`. It doesn't look substantially different from how it's used in the OTLP receiver, but it should help illustrate how this might look.